enhancements/kubelet/kubelet-node-sizing.md

```diff
@@ -0,0 +1,212 @@
+  creationTimestamp: null
+  name: testcluster1
+  autoNodeSizing: true
+networking:
+  clusterNetwork:
+  - cidr: 10.128.0.0/14
+    hostPrefix: 23
+  machineNetwork:
+  - cidr: 10.0.0.0/16
+  networkType: OpenShiftSDN
+  serviceNetwork:
+  - 172.30.0.0/16
+platform:
+  gcp:
+    projectID: openshift-gce-devel
+    region: asia-south1
+publish: External
+pullSecret: </snip>
+```
+
+### Kubelet systemd service
+
+```
+[Unit]
+Description=Kubernetes Kubelet
+Wants=rpc-statd.service network-online.target crio.service
+After=network-online.target crio.service
+
+[Service]
+Type=notify
+ExecStartPre=/bin/mkdir --parents /etc/kubernetes/manifests
+ExecStartPre=/bin/rm -f /var/lib/kubelet/cpu_manager_state
+ExecStartPre=/usr/local/bin/node-sizing.sh
+EnvironmentFile=/etc/os-release
+EnvironmentFile=-/etc/kubernetes/kubelet-workaround
+EnvironmentFile=-/etc/kubernetes/kubelet-env
+EnvironmentFile=/etc/kubernetes/node-sizing-env
+
+ExecStart=/usr/bin/hyperkube \
+    kubelet \
+      --config=/etc/kubernetes/kubelet.conf \
+      --bootstrap-kubeconfig=/etc/kubernetes/kubeconfig \
+      --kubeconfig=/var/lib/kubelet/kubeconfig \
+      --container-runtime=remote \
+      --container-runtime-endpoint=/var/run/crio/crio.sock \
+      --runtime-cgroups=/system.slice/crio.service \
+      --node-labels=node-role.kubernetes.io/worker,node.openshift.io/os_id=${ID} \
+      --node-ip=${KUBELET_NODE_IP} \
+      --minimum-container-ttl-duration=6m0s \
+      --volume-plugin-dir=/etc/kubernetes/kubelet-plugins/volume/exec \
+      --cloud-provider=gce \
+      --cloud-config=/etc/kubernetes/cloud.conf \
+      --pod-infra-container-image=quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:9519ae9a0a3e262e311c7f12a08adb2568e29e1576d2c6c229fd5d355c551d4b \
+      --system-reserved=cpu=${CPU_RESERVED},memory=${MEMORY_RESERVED},ephemeral-storage=${EPHEMERAL_RESERVED}
+      --v=${KUBELET_LOG_LEVEL}
+
+Restart=always
+RestartSec=10
+
+[Install]
+WantedBy=multi-user.target
+```
+
+Node sizing values will be calculated by the script `/usr/local/bin/node-sizing.sh`. This script will set the environment variables `CPU_RESERVED`, `MEMORY_RESERVED` and `EPHEMERAL_RESERVED` in `/etc/kubernetes/node-sizing-env` which will be used for setting the kubelet parameter `system-reserved`
+
+The script `/usr/local/bin/node-sizing.sh` will be executed using `ExecStartPre`
+
+#### Node sizing script
+
+Irrespective of `auto node sizing` is enabled or not, MCO will always place the script file `/usr/local/bin/node-sizing.sh` on the node.
+
+However the content of the script file will vary depending upon whether the `auto node sizing` is enabled or disabled. When the `auto node sizing` is disabled, `/usr/local/bin/node-sizing.sh` will set variables `CPU_RESERVED`, `MEMORY_RESERVED` and `EPHEMERAL_RESERVED` to their existing default values, but when the `auto node sizing` is enabled the script will dynamically select the optimal value based on the resources available on the node. 
```

Comment:
What is the plan for how the MCO will know whether auto node sizing is enabled? Is this something that we will allow the user to change as a day-2 operation?